### PR TITLE
plugin-host: add complete lifecycle disposal

### LINKED ISSUE #562: Add complete PluginManager and PluginLoaderService disposal
## Summary

`PluginManager` owns the JavaScript runtime, machine subscriptions, plugin instances, bridge state, asynchronous request registries, and an event stream, but it has no complete disposal contract. `PluginLoaderService` likewise has no lifecycle method that shuts the manager down.

This leaves Decaid unable to deterministically release the plugin subsystem during service replacement, tests, application shutdown paths, or future runtime restarts. It also makes it difficult to prove that the memory growth investigated in #519 is bounded once the native engine leaks are repaired.

Implement a terminal, idempotent lifecycle for the entire plugin subsystem.

Audit baseline: Decaid commit `61346e23c2e42234d6672abc36da33c9c2947edc`, reviewed on 2026-08-05.

## Related work

- #519 tracks the observed iOS memory growth and jetsam termination.
- #561 owns the detailed cleanup contracts for timers, fetches, plugin HTTP requests, Decent proxy requests, and other pending asynchronous operations.
- The JavaScriptCore backend also requires an engine-level ownership/disposal repair in `tadelv/dart_js`. GitHub Issues are currently disabled in that repository, so that ticket could not yet be created.

This issue owns the top-level lifecycle orchestration. #561 owns the individual pending-resource implementations that this disposal path must invoke.

## Current ownership

`PluginManager` currently owns or controls:

- `JavascriptRuntime js`;
- all loaded `PluginRuntime` records;
- per-plugin Decent proxy bridge tokens;
- the plugin event broadcast `StreamController`;
- the active `De1Controller` reference;
- the DE1 connection subscription;
- the machine snapshot subscription;
- JavaScript bridge registrations;
- plugin bootstrap globals and loaded plugin objects;
- pending plugin HTTP response state;
- fetch and proxy work initiated by JavaScript;
- any shared network client introduced by #561;
- plugin-owned timers and pending promises introduced/managed by #561.

`PluginLoaderService` owns the manager as part of the application service graph but does not expose a matching shutdown method.

### 1. No `PluginManager.dispose()`

The manager cannot currently:

- stop accepting plugin loads or events;
- await subscription cancellation;
- unload all plugins;
- close its event stream;
- cancel/reject pending plugin work;
- remove message channels and bridge state;
- dispose the JavaScript runtime;
- close a shared HTTP client;
- clear ownership maps;
- guarantee that later API calls fail predictably.

### 2. No `PluginLoaderService.dispose()`

The service layer cannot shut down the manager it created. Any application lifecycle or test teardown must rely on garbage collection and partial object loss rather than an explicit contract.

### 3. Controller replacement can overlap subscriptions

The `de1Controller` setter starts cancellation of the old subscriptions asynchronously but does not await completion before subscribing to the new controller. During a controller swap, old and new streams may overlap temporarily and old events may be delivered into the current plugin generation.

A disposed or superseded manager also needs protection against late events from a subscription whose cancellation completed after replacement began.

### 4. Plugin unload is not a complete teardown boundary

`unloadPlugin()` removes Dart-side records and evaluates JavaScript cleanup, but cleanup failures or asynchronous plugin work can outlive the unload call. The manager has no owner-level mechanism to ensure all plugin resources are cancelled before the plugin is considered gone.

The detailed timer/request cleanup is tracked by #561, but this issue must ensure it is always invoked in the correct order.

### 5. No terminal lifecycle state

Public methods can still be called while or after teardown. Without an explicit state, event dispatch, plugin loading, bridge completions, and runtime evaluation may race with released resources.

## Required lifecycle

Introduce an explicit manager lifecycle, for example:

```text
active -> disposing -> disposed
```

### While active

- Normal plugin loading, unloading, dispatch, and controller attachment are permitted.

### When disposal begins

- Transition to `disposing` exactly once.
- Reject or ignore new plugin loads, event dispatches, controller attachments, and bridge requests.
- Ensure concurrent calls to `dispose()` await the same teardown operation.

### Disposal order

Use an order that prevents new work from being created while existing work is removed. A suitable sequence is:

1. Mark the manager as disposing and stop accepting new work.
2. Detach controller references and await cancellation of DE1/snapshot subscriptions.
3. Prevent queued/late subscription callbacks from dispatching events.
4. Unload every loaded plugin, continuing cleanup even if one plugin fails.
5. Invoke the plugin-owned timer and pending-operation cancellation hooks from #561.
6. Reject any remaining bridge promises and remove bridge/channel registrations.
7. Close the plugin event `StreamController`.
8. Close the manager-owned shared `HttpClient`, once introduced by #561.
9. Dispose the JavaScript runtime.
10. Clear plugin maps, bridge tokens, pending registries, and controller references.
11. Mark the manager disposed.

Every cleanup stage should use `try/finally` or aggregated error handling so one failure does not skip later teardown.

### Post-disposal behavior

After disposal:

- `loadPlugin`, `dispatchEvent`, controller attachment, and similar mutating operations must fail with a clear `StateError` or follow another explicitly documented no-op contract;
- late JavaScript/network/subscription completions must be ignored without recreating state;
- repeated `dispose()` calls must be safe and await the already completed teardown;
- no method may evaluate JavaScript through a disposed runtime.

## Controller attachment redesign

The current synchronous property setter is not suitable for an operation that requires awaited cancellation.

Replace or supplement it with an asynchronous method such as:

```dart
Future<void> attachDe1Controller(De1Controller? controller)
```

or serialize controller changes internally.

Required behavior:

- await cancellation of previous subscriptions before treating the new controller as active;
- assign a monotonically increasing attachment/generation token;
- discard events from older generations;
- handle rapid successive controller changes deterministically;
- reject attachment once disposal begins;
- avoid duplicate snapshot subscriptions.

If the existing setter must remain for compatibility, it should delegate into a serialized lifecycle mechanism rather than launching untracked asynchronous cancellation.

## Plugin unload requirements

Top-level unload should:

- mark the plugin generation as stopping before invoking JavaScript;
- stop delivering new events to that generation;
- cancel plugin-owned timers and pending operations through #561;
- invoke `onUnload` with a bounded, documented contract;
- delete the JavaScript plugin registry entry in `finally`, even when `onUnload` throws;
- remove bridge tokens and owner registries;
- mark the Dart runtime record disposed only after host-owned cleanup has run;
- ignore late completions carrying the old generation ID.

A failure unloading one plugin must not prevent disposal of the remaining plugins.

## `PluginLoaderService` requirements

Add an idempotent `Future<void> dispose()` that:

- stops any loader-owned work;
- delegates to `PluginManager.dispose()`;
- prevents future plugin loading through the service;
- can be called safely by tests and the application's service teardown.

Wire this method into the actual application/service lifecycle where the loader service is replaced or shut down. Avoid adding a method that is never called outside tests.

## Diagnostics

Expose enough state for tests and beta diagnostics to verify teardown:

- lifecycle state;
- loaded plugin count;
- active subscription count or attachment generation;
- pending-operation counts supplied by #561;
- active timer count supplied by #561;
- JavaScript runtime disposed state where available.

Production logs should identify teardown failures without escalating ordinary third-party plugin `onUnload` failures into unrelated crash signals.

## Scope boundaries

This issue includes top-level Decaid plugin subsystem lifecycle and orchestration.

It does **not** include:

- repairing JavaScriptCore native ownership inside `tadelv/dart_js`;
- implementing each timer/network pending-operation registry in detail — that is #561;
- unifying JavaScript engines;
- redesigning per-event JavaScript dispatch;
- adding new plugin host capabilities;
- requiring real-device memory validation in CI where no representative iOS hardware is available.

The implementation should remain useful whether Decaid continues using JavaScriptCore on Apple or later switches to QuickJS.

## Acceptance criteria

- [ ] `PluginManager` has an idempotent asynchronous `dispose()` method.
- [ ] Concurrent calls to `dispose()` share one teardown and do not double-release resources.
- [ ] Disposal stops new plugin loads, controller attachments, and event dispatch before releasing resources.
- [ ] DE1 and snapshot subscriptions are awaited and no events are delivered after disposal.
- [ ] Controller replacement cannot leave overlapping active subscription generations.
- [ ] All plugins are unloaded even if one plugin's cleanup fails.
- [ ] JavaScript plugin registry entries are deleted even when `onUnload()` throws.
- [ ] Plugin-owned resources are cancelled through the cleanup contracts from #561.
- [ ] The event stream controller and shared HTTP client are closed exactly once.
- [ ] The JavaScript runtime is disposed exactly once and only after bridge/plugin work is stopped.
- [ ] All ownership maps and controller references are empty after disposal.
- [ ] Late completions from old plugin or controller generations are ignored and do not recreate state.
- [ ] Public mutating methods fail predictably after disposal.
- [ ] `PluginLoaderService` exposes and uses an idempotent `dispose()` method.
- [ ] The application/service teardown actually calls the loader/manager disposal path.

### Manager lifecycle

- create and dispose an empty manager;
- dispose a manager with multiple loaded plugins;
- call `dispose()` repeatedly and concurrently;
- attempt load, dispatch, and controller attachment while disposing and after disposal;
- verify the event stream is closed;
- verify the runtime receives exactly one disposal call.

### Subscription lifecycle

- attach a controller, replace it, and verify only the latest generation can dispatch;
- rapidly replace controllers multiple times;
- dispose while subscription cancellation is pending;
- deliver a simulated late event from an old generation and verify it is ignored;
- assert no active subscription remains after teardown.

### Plugin teardown

- unload a normal plugin;
- unload a plugin whose `onUnload()` throws;
- dispose multiple plugins where one unload fails;
- unload/reload the same plugin and verify late work from the old generation is ignored;
- verify all JavaScript registry entries and Dart ownership records are removed.

### Integration with #561

- dispose with armed timers;
- dispose with pending fetch, proxy, and plugin HTTP requests;
- verify every pending resource is cancelled/rejected and all counters return to zero.

### Repeated lifecycle stress

Run repeated manager create/load/unload/dispose cycles and verify:

- no listener/subscription growth;
- no retained `PluginRuntime` records;
- no pending-operation or timer growth;
- no post-disposal JavaScript calls.

## Device validation

CI should provide deterministic lifecycle and resource-invariant coverage. Full iOS memory validation can continue through the existing beta-user workflow for #519:

- compare the current build with the disposal and engine fixes;
- run the same plugin set and machine workload;
- confirm memory reaches a bounded plateau rather than only growing more slowly;
- confirm no plugin events or network completions occur after teardown.

### PR DESCRIPTION
**Pull request by ODevStudio**
## Summary

- Problem: PluginManager and PluginLoaderService had no terminal lifecycle, leaving subscriptions, timers, pending operations, host bridges, and the JavaScript runtime without complete teardown.
- Why it matters: stale callbacks could reach disposed or reloaded plugin generations, while process teardown leaked plugin-owned resources.
- What changed: added idempotent lifecycle disposal, serialized DE1 attachment, generation-guarded callbacks, complete unload cleanup, loader teardown, and detached-only application wiring.
- What did NOT change (scope boundary): public plugin events/APIs and native JavaScriptCore/QuickJS ownership inside flutter_js.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [x] Refactor required for the fix
- [x] Docs
- [ ] Security hardening
- [ ] Chore / infra
- [ ] Plugin (DYE2 or bundled skin)

## Scope (select all touched areas)

- [ ] BLE transport / device comms
- [ ] REST API / handlers
- [ ] WebSocket API
- [ ] Machine state / shot logic
- [ ] Scale / weight / flow
- [ ] Profiles / beans / grinders / workflows
- [ ] WebUI skins
- [x] Plugins / JS runtime
- [x] UI / Flutter widgets
- [ ] Storage / Drift database
- [ ] CI / build / infra
- [x] Docs / specs

## Linked Issues

- Closes #562
- Related #561

## Root Cause (if bug fix)

- Root cause: plugin services had no terminal ownership boundary, and controller replacement cancelled subscriptions without awaiting completion.
- Missing detection or guardrail: no lifecycle-specific tests covered concurrent disposal, stale generations, event-stream completion, or outer application teardown.
- Contributing context (if known): #561 added bounded async ownership but intentionally left subsystem disposal for this follow-up.

## Regression Test Plan (if bug fix or refactor)

- Coverage level that should have caught this:
  - [x] Unit test
  - [ ] Integration test (mock transport edge)
  - [ ] End-to-end test (simulate=1 + curl/websocat)
  - [ ] Existing coverage already sufficient
- Target test or file: test/plugins/plugin_manager_lifecycle_test.dart, test/plugin_loader_service_lifecycle_test.dart, test/app_lifecycle_observer_test.dart
- Scenario the test should lock in: teardown is shared, terminal, complete, generation-safe, and invoked only on process detachment.
- If no new test added, why not: N/A

## Documentation Obligations (required)

- [ ] API spec updated: assets/api/rest_v1.yml or assets/api/websocket_v1.yml (if REST/WebSocket changed)
- [ ] API docs updated: doc/Api.md (if user-facing endpoint changed)
- [ ] Plugin docs updated: doc/Plugins.md (if events/API changed)
- [ ] Skin docs updated: doc/Skins.md (if skin behavior changed)
- [ ] Profile docs updated: doc/Profiles.md (if profile handling changed)
- [ ] Device docs updated: doc/DeviceManagement.md (if device flows changed)
- [x] N/A - no public docs, events, or API contracts changed

## Security Impact (required)

- New or changed REST endpoints? No
- New or changed WebSocket topics? No
- New or changed network calls? No
- BLE/USB surface changed? No
- File system access changed? No
- Plugin sandbox boundary changed? Yes
- If any Yes, explain risk and mitigation: host messages now carry and validate plugin generations so stale deferred work cannot target a reloaded plugin or recreate state after teardown.

## User-Visible Changes

None.

### Local gates (run before pushing)

- [x] dart format lib test - no remaining changes
- [x] flutter analyze - clean (no new warnings)
- [ ] flutter test - full run completed with 2,821 passes and 9 unrelated baseline/environment failures: eight existing Windows _copyDirectory failures and one absent generated bundled-skin manifest
- [ ] ./scripts/fetch_dye2_plugin.sh - DYE2 plugin installed into assets/plugins/dye2.reaplugin/

### Manual verification (if applicable)

- OS / platform tested: Windows
- Simulated devices? (simulate=1): No
- Real hardware? (DE1/Bengle/scale): No
- What you personally verified and how: 10 focused lifecycle tests and 25 affected plugin ownership tests passed; flutter analyze reported no issues.
- Edge cases checked: empty/error-path disposal, concurrent disposal, throwing onUnload, stale generations, controller replacement, pending cancellation, event completion, and detached application teardown.
- What you did not verify: real hardware, native JavaScriptCore/QuickJS allocation behavior, and generated bundled-skin assets.

### Evidence

- [x] Test output (failing before + passing after)
- [ ] Log snippets
- [ ] Screenshot / recording (UI changes)
- [ ] curl / websocat output (API changes)

## Compatibility & Migration

- Backward compatible? Yes
- Config / env changes needed? No
- Database migration needed? No
- If any No or Yes, explain exact steps: No migration or configuration changes are required.

## Risks & Mitigations

- Risk: asynchronous callbacks race with plugin unload or process detachment.
  - Mitigation: lifecycle guards, monotonically increasing generations, serialized subscription replacement, and focused race/teardown tests.

### CODE REVIEW

**Review by tadelv — CHANGES_REQUESTED** [review]
The Decaid-side lifecycle work looks sound overall and I did not find another main-use-case blocker, but this PR needs to be validated against the `dart_js` fix it is intended to complement before merging.

PR #582 is still locked to `flutter_js` / `dart_js` commit `c7cde7b5c9c6ad7013157f296a7320835aac6cdb`, which is the pre-fix baseline of `tadelv/dart_js#4`. That PR has since merged the JavaScriptCore ownership, bridge, and terminal-disposal fixes required for #519.

The current #582 implementation also directly accesses `JavascriptRuntime.channelFunctionsRegistered` during manager disposal. The merged `dart_js` runtime no longer exposes that process-global registry: channel/runtime cleanup is now owned by `JavascriptRuntime.dispose()`. The `_CountingRuntime.setupBridge` test override also uses the old runtime API signature.

Please update/re-resolve `flutter_js` to the merged `dart_js` revision and make #582 compile and pass its lifecycle/plugin tests against that version. In particular, remove the direct static channel-registry cleanup and let `js.dispose()` own engine/channel teardown, and update the test runtime wrapper to the current bridge API.

This matters for #519: #577/#561 handles Decaid's pending async ownership, this PR handles the top-level plugin lifecycle, and `dart_js#4` handles the engine-level native ownership. We should verify those three pieces together rather than merge #582 based on CI against the exact engine revision known to have the #519 defects.

Once that dependency integration is updated and green, I don't currently see another blocker in the main lifecycle path.

**Comment by ODevStudio** [comment]
Addressed the requested dart_js integration in 95567e88 and merged the latest main in 08a10b81. The branch remains pinned to dart_js merge commit 5bdb99666e84beb90f01f1b89c2949a36c7d32b0, delegates engine/channel teardown to JavascriptRuntime.dispose(), and uses the current setupBridge API. Post-merge verification: all 54 plugin/lifecycle tests pass, flutter analyze is clean, and all 4 GitHub checks pass. @tadelv, please re-review when convenient.

File: lib/main.dart [hunk 1]
```diff
@@ -743,6 +749,11 @@ class AppLifecycleObserver with WidgetsBindingObserver {
 
   @override
   void didChangeAppLifecycleState(AppLifecycleState state) {
+    if (state == AppLifecycleState.detached && !_detaching) {
```
Comment (tadelv) [inline lib/main.dart]: `detached` is not a cross-platform process-shutdown hook. Flutter documents that this state is re-entered after startup only on iOS/Android/web; desktop platforms do not re-enter it, while Decaid supports macOS/Linux/Windows. As written, normal desktop exit therefore never calls `pluginLoaderService.dispose()`, so the top-level ownership boundary added by this PR is still unused on desktop (including macOS, which matters to the #519 investigation). Please also wire disposal to Decaid's actual desktop exit path. `WidgetsBindingObserver.didRequestAppExit()` is awaitable for the desktop platforms Flutter currently supports there (macOS/Linux); keep `detached` for mobile and add the appropriate shutdown handling/fallback for the remaining desktop path, with a test that proves desktop exit reaches disposal.

**Review by tadelv — CHANGES_REQUESTED** [review]
The Decaid-side manager/loader lifecycle looks sound after the dart_js integration update, and the current PR checks are green. I found one remaining application-lifecycle blocker: the terminal teardown is wired only to `AppLifecycleState.detached`, which is not a desktop shutdown notification. Please wire the loader disposal into the actual desktop exit path as well and cover that path with a focused test.

**Review by tadelv — APPROVED** [review]
